pkg/scheduler/queue/queue_test.go

```diff
@@ -63,10 +78,149 @@ func makeSchedulerRequest(tenantID string) *SchedulerRequest {
 		},
-		AdditionalQueueDimensions: randAdditionalQueueDimension(),
+		AdditionalQueueDimensions: additionalQueueDimensions,
+		EnqueueTime:               time.Now(),
 	}
 }
 
+// TestMultiDimensionalQueueFairnessSlowConsumerEffects emulates a simplified queue slowdown scenario
+// which the scheduler's additional queue dimensions features are intended to solve for.
+//
+// In this scenario, one category of queue item causes the queue consumer to slow down, introducing a
+// significant delay while the queue consumer processes it and before the consumer can dequeue the next item.
+// This emulates a situation where one of the query components - the ingesters or store-gateways - is under load.
+//
+// If queue items belonging to the slow category are in the same queue in front of the normal queue items,
+// the normal queue items must wait for all slow queue items to be cleared before they can be serviced.
+// In this way, the degraded performance of the slow query component equally degrades the performance of the
+// queries which *could* be serviced quickly, but are waiting behind the slow queries in the queue.
+//
+// With the additional queue dimensions enabled, the queues are split by which query component the query will utilize.
+// The queue broker then round-robins between the split queues, which has the effect of alternating between
+// dequeuing the slow queries and normal queries rather than blocking normal queries behind slow queries.
+func TestMultiDimensionalQueueFairnessSlowConsumerEffects(t *testing.T) {
+	promRegistry := prometheus.NewPedanticRegistry()
+
+	maxQueriersPerTenant := 0 // disable shuffle sharding
+	forgetQuerierDelay := time.Duration(0)
+	maxOutstandingRequestsPerTenant := 1000
+
+	totalRequests := 100
+	numTenants := 1
+	numProducers := 10
+	numConsumers := 1
+
+	normalQueueDimension := storeGatewayQueueDimension
+	slowConsumerLatency := 20 * time.Millisecond
+	slowConsumerQueueDimension := ingesterQueueDimension
+	normalQueueDimensionFunc := func() []string { return []string{normalQueueDimension} }
+	slowQueueDimensionFunc := func() []string { return []string{slowConsumerQueueDimension} }
+
+	additionalQueueDimensionsEnabledCases := []bool{false, true}
+	queueDurationTotals := map[bool]map[string]float64{
+		false: {normalQueueDimension: 0.0, slowConsumerQueueDimension: 0.0},
+		true:  {normalQueueDimension: 0.0, slowConsumerQueueDimension: 0.0},
+	}
+
+	for _, additionalQueueDimensionsEnabled := range additionalQueueDimensionsEnabledCases {
+
+		// production code uses a histogram for queue duration, but the counter is a more direct metric for test cases
+		queueDuration := promauto.With(promRegistry).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_query_scheduler_queue_duration_total_seconds",
+			Help: "Total time spent by requests in queue before getting picked up by a querier.",
+		}, []string{"additional_queue_dimensions"})
```

Comment:
Do we need a Prometheus metric at all here? I'm wondering if we could just have two atomic floats, one for normal requests and the other for slow requests, and accumulate the total duration in those

Reply:
sure, but isn't the prometheus metric just a more convenient interface for multiple atomic floats? It also allows us to reuse the same code for recording the observations as we use in the prod scheduler code.